firewall_test.go

```diff
@@ -0,0 +1,109 @@
+package libnetwork
+
+import (
+	"fmt"
+	"github.com/docker/libnetwork/iptables"
+	"github.com/docker/libnetwork/netlabel"
+	"github.com/docker/libnetwork/options"
+	"strings"
+	"testing"
+)
+
+const (
+	fwdChainName = "FORWARD"
+	usrChainName = "DOCKER-USER"
+)
+
+func verifyUserChain(enabled, insert bool) error {
+	output, err := iptables.Raw("-S", fwdChainName)
+	if err != nil {
+		return err
+	}
+	rules := strings.Split(string(output), "\n")
+	if !enabled {
+		if len(rules)-1 != 1 {
+			return fmt.Errorf("chain %v: unexpected rules len=%v, %v, ",
+				fwdChainName, len(rules), rules)
+		}
+		if _, err = iptables.Raw("-S", usrChainName); err == nil {
+			return fmt.Errorf("chain %v: created unexpectedly", usrChainName)
+		}
+		return nil
+	}
+	nRules := 2
+	if insert {
+		nRules++
+	}
+	if nRules != len(rules)-1 || !strings.Contains(rules[1], usrChainName) {
+		return fmt.Errorf("chain %v: unexpected rules len=%v, %v",
+			fwdChainName, len(rules), rules)
+	}
+
+	output, err = iptables.Raw("-S", usrChainName)
+	if err != nil {
+		return err
+	}
+	rules = strings.Split(string(output), "\n")
+	if len(rules)-1 != 2 {
+		return fmt.Errorf("chain %v: unexpected rules len=%v, %v",
+			usrChainName, len(rules), rules)
+	}
+	return nil
+}
+
+func testUserChain(t *testing.T, ctrl *controller, enabled, insert bool) {
+	defer func() {
+		_, err := iptables.Raw("-F", fwdChainName)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_ = iptables.RemoveExistingChain(usrChainName, "")
+	}()
+	// init. condition, FORWARD chain empty
+	// DOCKER-USER not exist
+	err := verifyUserChain(false, insert)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if insert {
+		_, err = iptables.Raw("-A", fwdChainName, "-j", "DROP")
```

Comment:
the reason to have insert is that in case there is already a rule in FORWARD chain, DOCKER-CHAIN rule would be placed ahead of that rule when iptable=true; if iptable=false, it is moot because we do add DOCKER-CHAIN rule to FORWARD chain.

Reply:
Right, so the thing that worries me a bit in the test, is that we're creating a new code-path _just_ for the test. The production code does not have check wether or not this should be called. (The production code would (currently) call `setupArrangeUserFilterRule` and `arrangeUserFilterRule` when calling `controller.NewNetwork()`, which is not the code-path we're testing.

And because of `userChainOnce` and `ctrl` being global, we can't test different permutations of iptables being enabled/disabled etc.